sstables/sstable_set.cc

```diff
@@ -769,6 +770,393 @@ sstable_set_impl::create_single_key_sstable_reader(
+                    }
+
+                    // One of the existing readers won with the galloping reader,
+                    // or there is a yet unselected reader which possibly has a smaller position.
+                    // In either case we exit the galloping mode.
+
+                    _peeked_readers.emplace_back(_galloping_reader);
+                    boost::range::push_heap(_peeked_readers, _peeked_cmp);
+                }
+            } else {
+                // See comment in `peek_reader`.
+                _halted_readers.push_back(_galloping_reader);
+                boost::range::push_heap(_halted_readers, _halted_reader_cmp);
+            }
+
+            // The galloping reader has either been removed, halted, or lost with the other readers.
+            // Proceed with the normal path.
+            _galloping_reader = {};
+            _gallop_mode_hits = 0;
+            return (*this)(timeout);
+        });
+    }
+
+public:
+    // Assumes the readers returned from `reader_queue` return fragments in the same, single partition.
+    clustering_order_reader_merger(
+            schema_ptr schema, reader_permit permit, dht::decorated_key key,
+            streamed_mutation::forwarding fwd_sm,
+            min_position_reader_queue reader_queue)
+        : _schema(std::move(schema)), _permit(std::move(permit)), _key(key)
+        , _cmp(*_schema)
+        , _reader_queue(reader_queue)
+        , _peeked_cmp(*_schema)
+        , _halted_reader_cmp(*_schema)
+        , _pr_end(fwd_sm == streamed_mutation::forwarding::yes
+                        ? position_in_partition::before_all_clustered_rows()
+                        : position_in_partition::after_all_clustered_rows())
+        , _should_return_partition_end(fwd_sm == streamed_mutation::forwarding::no) {}
+
+    // We assume that operator() is called sequentially and that the caller doesn't use the batch
+    // returned by the previous operator() call after calling operator() again
+    // (the data from the previous batch is destroyed).
+    future<mutation_fragment_batch> operator()(db::timeout_clock::time_point timeout) {
+        if (!_returned_partition_start) {
+            _current_batch.push_back(mutation_fragment(*_schema, _permit, partition_start(_key, {})));
+            _returned_partition_start = true;
+            return make_ready_future<mutation_fragment_batch>(_current_batch);
+        }
+
+        if (in_gallop_mode()) {
+            return peek_galloping_reader(timeout);
+        }
+
+        if (!_unpeeked_readers.empty()) {
+            return peek_readers(timeout).then([this, timeout] { return (*this)(timeout); });
+        }
+
+        auto next_peeked_pos = _peeked_readers.empty() ? _pr_end : _peeked_readers.front()->reader.peek_buffer().position();
+        // There might be queued readers with min_position <= next_peeked_pos:
+        if (!_reader_queue.empty(next_peeked_pos)) {
+            auto rs = _reader_queue.pop(next_peeked_pos, timeout);
```

Comment:
Won't this effectively pull in all remaining sstables as soon as we run out in the middle? Or will the queue reader just return the minimum amount of readers possible like the LCS selector?

Reply:
It will be pulling sstables until it sees that the next sstable in the queue has a greater min_position than the fragment with the smallest position among fragments returned by readers that were already pulled. So it tries to pull the minimum amount of readers based on the information we have.

Reply:
Ok, so it works like the LCS selector, which also always returns the next smallest group.